proptest/src/num/float_samplers.rs

```diff
@@ -0,0 +1,338 @@
+//-
+// Copyright 2022 Michael Wright
+//
+// Licensed under the Apache License, Version 2.0 <LICENSE-APACHE or
+// http://www.apache.org/licenses/LICENSE-2.0> or the MIT license
+// <LICENSE-MIT or http://opensource.org/licenses/MIT>, at your
+// option. This file may not be copied, modified, or distributed
+// except according to those terms.
+
+//! Alternative uniform float samplers because the ones provided by the rand crate are prone
+//! to overflow. The samplers work by uniformly selecting from a set of equally spaced values in
+//! the interval and the included bounds. Selection is slighly biased towards the bounds.
+
+pub(crate) use self::f32::F32U;
+pub(crate) use self::f64::F64U;
+
+macro_rules! float_sampler {
+    ($typ: ident, $int_typ: ident, $wrapper: ident) => {
+        pub mod $typ {
+            use rand::prelude::*;
+            use rand::distributions::uniform::{
+                SampleBorrow, SampleUniform, Uniform, UniformSampler,
+            };
+
+            #[must_use]
+            // Returns the previous float value. In other words the greatest value representable
+            // as a float such that `last_float(a) < a`. `-0.` is treated as `0.`.
+            fn last_float(a: $typ) -> $typ {
+                debug_assert!(a.is_finite() && a > $typ::MIN, "`last_float` invalid input: {}", a);
+                if a == (0.) {
+                    -$typ::from_bits(1)
+                } else if a < 0. {
+                    $typ::from_bits(a.to_bits() + 1)
+                } else {
+                    $typ::from_bits(a.to_bits() - 1)
+                }
+            }
+
+            #[must_use]
+            fn last_gap(a: $typ) -> $typ {
+                debug_assert!(a.is_finite() && a > $typ::MIN, "`last_float` invalid input: {}", a);
+                a - last_float(a)
+            }
+
+            #[derive(Copy, Clone, Debug)]
+            pub(crate) struct $wrapper($typ);
+
+            impl From<$typ> for $wrapper {
+                fn from(x: $typ) -> Self {
+                    $wrapper(x)
+                }
+            }
+            impl From<$wrapper> for $typ {
+                fn from(x: $wrapper) -> Self {
+                    x.0
+                }
+            }
+
+            #[derive(Clone, Copy, Debug)]
+            pub(crate) struct FloatUniform {
+                uniform: Uniform<$int_typ>,
+                values: SampleValueCollection,
+            }
+
+            impl UniformSampler for FloatUniform {
+
+                type X = $wrapper;
+
+                fn new<B1, B2>(low: B1, high: B2) -> Self
+                where
+                    B1: SampleBorrow<Self::X> + Sized,
+                    B2: SampleBorrow<Self::X> + Sized,
+                {
+                    let low = low.borrow().0;
+                    let high = high.borrow().0;
+
+                    let values = SampleValueCollection::new_inclusive(low, last_float(high));
```

Comment:
if someone were to provide a range between low and high, wouldn't the expectation be that that range sets the bounds for the sampling? by funneling high through `last_float`, it's decreasing the high bounds by 1 bit of magnitude. I'm sure there's a reason for this that i'm not understanding, could you help me better understand? Possible question here is, what does "new_inclusive" imply that "new" does not?

Reply:
this trait is from Rand so we don't have much semantic control over. That said, UniformSampler provides 2 constructors to account for inclusive and exclusive upper bounds.

`new` samples in a range with exclusive upper bound `[low, high)` or `low..high` in rust range syntax

`new_inclusive` samples in a range with inclusive upper bound `[low, high]` or `low..=high` in rust range syntax

https://docs.rs/rand/latest/rand/distributions/uniform/trait.UniformSampler.html#tymethod.new

Reply:
my comment doesn't make a whole lot of sense since you're specifically referring to `SampleValueCollection`, not the `UniformSampler` methods. 

I like that our internal impls follow the same naming convention which is why I brought up UniformSampler.

Specifically for why we only have `new_inclusive` -- the previous revision of this PR had both `new` and `new_inclusive` but there was a slight bug where we ended up doubling up on calls to `last_float` so i suggested condensing to just a single `new_inclusive` that was unambiguous and just implies the range includes the value passed in as the second parameter

Reply:
ah, thank you that explains a lot